### PR TITLE
Initial trigger log support

### PR DESCRIPTION
**Pull request by rick-a-lane-ii**
Initial foundation of core working components for streaming trigger logs.
- Intercepts `stdout` and `stderr` in the entrypointer.
- Adds logging to the metadata API.
- Adds the log service client.
- Determines the logging context from the available metadata information.
- Adds minimal testing of the metadata API logging.

A lot of improvements to be made down the line...

### CODE REVIEW

File: pkg/metadataapi/server/api/server.go
Comment (impl) [inline pkg/metadataapi/server/api/server.go:40]: Wouldn't we want this to be `/logs/{logId}/messages`?

File: pkg/metadataapi/server/api/logs.go
Comment (impl) [inline pkg/metadataapi/server/api/logs.go:19]: It feels weird to me to accept a protobuf message over a non-protobuf protocol, especially in the case where it's encoding some standard REST-ish headers (and see the other comment about including the log ID in the path...). I think this handler should just accept bytes with a `Content-Type` of `application/octet-stream`, similar to `PutOutput`. What do you think?

File: pkg/metadataapi/server/api/logs.go
Comment (impl) [inline pkg/metadataapi/server/api/logs.go:38]: Similar thought here. I think we should have a really simple envelope struct with only a `Name` field, and in the PLS-specific manager we can translate it to the proper protobuf request.

File: pkg/model/log.go
Comment (impl) [inline pkg/model/log.go:9]: Along the same lines as the HTTP endpoints, I think it might be a good idea to put in a bit more abstraction here, and also to distance the managers from the HTTP endpoint in naming to make it clear that they are standalone. I'm thinking something like:

```go
type Log struct {
  ID string
  Name string
}

type LogMessage struct {
  Log *Log
  Payload []byte
}

type LogManager interface {
  Create(ctx context.Context, name string) (*Log, error)
  AddMessage(ctx context.Context, logID string, payload []byte) (*LogMessage, error)
}
```

This would make this manager feel consistent with the other managers in this package.

File: pkg/operator/obj/webhooktriggerdeps.go
Comment (impl) [inline pkg/operator/obj/webhooktriggerdeps.go:264]: Do we need to add this? I don't think the trigger sync ID can change without the entire object being replaced, can it?

File: pkg/operator/obj/webhooktriggerdeps.go
Comment (impl) [inline pkg/operator/obj/webhooktriggerdeps.go:254]: Can we just fix this upstream in the tenant code or API code?

File: pkg/metadataapi/server/middleware/auth.go
Comment (impl) [inline pkg/metadataapi/server/middleware/auth.go:177]: If the log service client or log context ends up empty, we might not want to set this manager at all (leave it with the reject manager).

File: cmd/relay-runtime-tools/runner.go
Comment (impl) [inline cmd/relay-runtime-tools/runner.go:71]: We should exit after this, right? Otherwise it will just panic below.

File: pkg/model/model.go
Comment (impl) [inline pkg/model/model.go:17]: Can we put some comments around these `managed` labels to indicate that they should be pulled out of this project once we have this fixed on the API side?

**Review by impl — COMMENTED** [review]
Overall, the structure of this and tests look good from here! I have a couple of smallish thoughts, mostly around naming/envelope abstraction. Let me know what you think.

File: pkg/metadataapi/server/api/server.go
Comment (rick-a-lane-ii) [reply pkg/metadataapi/server/api/server.go:40]: I don't believe so. The log ID is embedded in the log object we're sending and is a fairly critical part of it. Adding the {logId} here feels like we're trying to maintain a REST API structure just for the sake of it. I don't like situations where we have the same data duplicated in two different places for no reason. That directly leads to (at least) two fundamental software engineering problems.

I believe this is a closer representation to the streaming nature here, and I don't see a need to really separate this out artificially. Everything about it feels redundant.

Originally the API signature included the {logId} and used it's own log structure that did not include the log ID and only the message. And I think that makes sense if you want to include the {logId} in the API. But if we're going to reuse the PLS structure (which I think is a better direction) and I believe what we called for in the spec, it makes more sense to have these APIs represent the gRPC functionality a bit closer.

I don't really think it matters either way, and I have a hard time justifying adding the {logId} in for no real reason.

File: pkg/model/log.go
Comment (rick-a-lane-ii) [reply pkg/model/log.go:9]: This was the way it was originally. I have mixed feelings about this, and it's going to be a repeating concern.

I ended up referencing the PLS objects directly in most of the code. Partly because we (essentially) called for it in the spec.

It's also a uniformly bad idea to duplicate models in multiple places rather than having one centralized source of truth.
And I believe we're going to be better off adding more abstraction and more implementations to the PLS, rather than focusing this within Relay core. Same argument applies to anyone wanting to used an entirely different PLS implementation; it feels like a better option to insist on the PLS contract rather than a vaguely compliant, generic option.

It's a valid argument to somewhat separate just this particular Manager interface. But it comes with all of the standard detractions to duplication.

File: pkg/model/model.go
Comment (rick-a-lane-ii) [reply pkg/model/model.go:17]: Let me see how easy it is to just fix this instead. I'm not sure I want these to remain in the API itself, but this definitely isn't the place for them either.

File: cmd/relay-runtime-tools/runner.go
Comment (rick-a-lane-ii) [reply cmd/relay-runtime-tools/runner.go:71]: That was originally there to ensure we do not fail (or timeout) in the existing integration tests and development environment if the PLS is not configured. But I believe I've added all the handling for that downstream now, so I'll verify that.

Additionally, the idea right now is to continue with the task at hand (no pun intended) even if failures occur during the logging, and to continue to write to stdout/stderr in those cases.

This could use a safety check below, though.

Long term, I'm not entirely sure how we want to handle errors within this code.

Comment (rick-a-lane-ii) [reply cmd/relay-runtime-tools/runner.go:71]: Obviously we'll need to ensure we handle transient errors properly and thoroughly, but in the case of unrecoverable errors, I don't think the answer it going to be interrupt the task if cannot log the output.

File: pkg/operator/obj/webhooktriggerdeps.go
Comment (rick-a-lane-ii) [reply pkg/operator/obj/webhooktriggerdeps.go:264]: It was accomplishing nothing before, as this data wasn't being set.
I'm not sure any of these can justifiable change without it referencing a different object.

Comment (rick-a-lane-ii) [reply pkg/operator/obj/webhooktriggerdeps.go:264]: The argument for including the trigger sync ID is the same as these other two items. They shouldn't change, but if they do, we should probably update the token. Although I'd probably be fine with removing this altogether as well.

File: pkg/metadataapi/opt/config.go
Comment (rick-a-lane-ii) [reply pkg/metadataapi/opt/config.go:182]: Ahh yes, a leftover from changing strategy.

File: pkg/metadataapi/server/api/server.go
Comment (rick-a-lane-ii) [reply pkg/metadataapi/server/api/server.go:40]: This is another instance where the benefits to using gRPC would be somewhat nullified (i.e. in this case, using a resource-based system anyway). I don't believe this feels like a "resource" per se. Obviously it could be viewed either way, but a "log" in this case does not feel like a resource. The log id feels like what it is: an index to identify the log.

File: pkg/model/log.go
Comment (rick-a-lane-ii) [reply pkg/model/log.go:9]: I believe we'll actually make this more explicit, not less. We're going to negate some of the explicit reasons for using gRPC if most of the code follows the existing interface design of the REST API and the Managers. Eventually, I'm leaning towards changing this to a "service" interface and removing this from the Managers altogether. The testing would be altered to focus more on verifying the API at that point, without necessarily adding an explicit "memory" option. It would likely move in the direction of implementing some microservices pattern rather than conform to the existing design at all. And this would still leave this to use the exact gRPC models. The generation and use of these from the proto file is part of the power of gRPC. As I said, this was the way it was coded originally, but it feels like duplicating the same content over and over again for no benefit. And also, as I've previously mentioned, I believe we're better off adding more options in the PLS functionality rather than here, if necessary (i.e a simpler and easier method to store logs rather than the BigQuery-dominated approach for the main path). This would require someone writing their own service to adhere to the interface defined by the PLS, but again, I believe that's an improvement over ambiguity and abstraction here.

File: pkg/metadataapi/server/api/logs.go
Comment (rick-a-lane-ii) [reply pkg/metadataapi/server/api/logs.go:38]: See below.

File: pkg/operator/obj/webhooktriggerdeps.go
Comment (rick-a-lane-ii) [reply pkg/operator/obj/webhooktriggerdeps.go:254]: So I avoided this for a few reasons. We're going to try to improve this handling on the API side. I don't like the options there either, as we're still working around the issue in a sense, but it is a better option regardless.

File: pkg/metadataapi/server/middleware/auth.go
Comment (rick-a-lane-ii) [reply pkg/metadataapi/server/middleware/auth.go:177]: Will do.

File: cmd/relay-runtime-tools/runner.go
Comment (rick-a-lane-ii) [reply cmd/relay-runtime-tools/runner.go:71]: Added a bit of necessary error handling based on this comment.

File: pkg/metadataapi/server/api/logs.go
Comment (rick-a-lane-ii) [reply pkg/metadataapi/server/api/logs.go:19]: On the surface I like this suggestion, but it would currently require this to be reflected to the Manager interface and the specific service interface, and that's moving further away from the direction we want to go in. The service interface shouldn't be receiving generic or abstract models.

Comment (rick-a-lane-ii) [reply pkg/metadataapi/server/api/logs.go:19]: I haven't seen anything that leads me to believe that sending the gRPC models over REST would be an issue. I feel like they are at least attempting to ensure that works properly.

File: pkg/model/model.go
Comment (kyleterry) [reply pkg/model/model.go:17]: I think these should be defined and godoc'd as constants in the controllers that use them, but that's just my 0.02.

**Review by kyleterry — COMMENTED** [review]
I've looked at this PR and it looks okay to me, but I'm going to wait for discourse between you and @impl to resolve before marking it as approved. This is looking great!

**Comment by rick-a-lane-ii** [comment]
Refactored the API and the LogManager interface to be more abstract. The only direct references or use of the PLS is in the client (i.e. relay-runtime-tools) and the end service. But it is still heavily based on protocol buffers, perhaps even more so now.

**Comment by rick-a-lane-ii** [comment]
Any none protocol buffer use of the log API can be layered on later, and adjustments made as necessary.

**Review by impl — APPROVED** [review]
LGTM!

**Pull request merged**
